Develop

IliaKanyshev opened the pull request:
add-remove-endpoint - Удаление фильмов и пользователей.
add-common-films - Функциональность «Общие фильмы».
add-most-populars - Вывод самых популярных фильмов по жанру и годам.
add-recommendations - Функциональность «Рекомендации».
add-director - Добавление режиссёров в фильмы.
add-search - Функциональность «Поиск».
add-reviews - Функциональность «Отзывы».
add-feed - Функциональность «Лента событий».

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

EndorpfinSplash committed:
Merge pull request #10 from IliaKanyshev/add-remove-endpoint

TZ-12 upgrade

OlegSavinIG committed:
directors added - почти все работает

надо отредактировать удаление режиссера

OlegSavinIG committed:
Update DirectorDbStorage.java

исправлено удаление

OlegSavinIG committed:
небольшие изменения

IliaKanyshev committed:
TZ-12 upgrade

EndorpfinSplash committed:
fix: director's film Ordering

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

IliaKanyshev committed:
TZ-12 upgrade

EndorpfinSplash committed:
Merge pull request #12 from IliaKanyshev/add-director

Add director

OlegSavinIG committed:
tests for directors and logs

TimurSurgut committed:
Добавлен Вывод самых популярных фильмов по жанру и годам.

IliaKanyshev committed:
feat: new method getFilmsByTitleOrDirector

EndorpfinSplash committed:
minor fixes
Final draft solution. not tested.
ReviewEntity for film. Tables and Controller add.

TimurSurgut committed:
Удалены лишние методы. Взамен них использована логика из FilmService.

TimurSurgut committed:
Добавлена Функциональность «Рекомендации».

TimurSurgut committed:
Подправил стиль в коде SQL

TimurSurgut committed:
Подправил стиль в коде SQL

OlegSavinIG committed:
tests for directors controller and logs

OlegSavinIG committed:
tests for directors controller and logs

OlegSavinIG committed:
common-films added

IliaKanyshev committed:
Merge pull request #15 from IliaKanyshev/add-common-films

Add common films

IliaKanyshev committed:
Clean up unused imports and checkstyle

IliaKanyshev committed:
Clean up unused imports and checkstyle

IliaKanyshev committed:
Merge remote-tracking branch 'origin/develop' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

IliaKanyshev committed:
Merge remote-tracking branch 'origin/add-most-populars' into add-most-populars

EndorpfinSplash committed:
Merge pull request #17 from IliaKanyshev/add-most-populars

Add most populars

IliaKanyshev committed:
Merge remote-tracking branch 'origin/develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java

IliaKanyshev committed:
fix: removed repeats in method getFilmsByTitleOrDirector

EndorpfinSplash committed:
Merge pull request #14 from IliaKanyshev/add-search

Add search PR

IliaKanyshev committed:
Merge remote-tracking branch 'origin/develop' into add-recommendations

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

IliaKanyshev committed:
feat: recommendations

EndorpfinSplash committed:
Merge pull request #18 from IliaKanyshev/add-recommendations

Add recommendations PR

Kumyshev committed:
add

Kumyshev committed:
refactor: классический вариант

EndorpfinSplash committed:
Business logic refactored

Kumyshev committed:
*

Kumyshev committed:
*

EndorpfinSplash committed:
test: checkstyle
 all passed

IliaKanyshev committed:
Merge pull request #19 from IliaKanyshev/add-reviews

Add reviews

IliaKanyshev committed:
Merge remote-tracking branch 'origin/develop' into add-feed

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java
#	src/main/resources/schema.sql

Kumyshev committed:
Merge remote-tracking branch 'origin/add-reviews' into add-feed

IliaKanyshev committed:
feat: recommendations

IliaKanyshev committed:
Merge remote-tracking branch 'origin/add-feed' into add-feed

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java
#	src/main/resources/schema.sql

Commit:
fix: save Review logging

Commit:
fix:  Review logging when exception happens

Commit:
fix:  Review logging when exception happens

Kumyshev committed:
*

Kumyshev committed:
Merge branch 'add-feed' of https://github.com/IliaKanyshev/java-filmorate into add-feed

Kumyshev committed:
Update FilmReviewDbStorage.java

IliaKanyshev committed:
refactor: removed some unused strings

IliaKanyshev committed:
Merge pull request #20 from IliaKanyshev/add-feed

Add feed PR

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java:
```diff
@@ -80,17 +76,41 @@ public Film getFilmById(Integer id) {
+                + "LEFT JOIN films f ON f.FILM_ID = df.FILM_ID "
+                + "LEFT JOIN likes l ON f.FILM_ID = l.FILM_ID "
+                + "LEFT JOIN MPA MPA on MPA.MPA_RATING_ID = f.MPA_RATING_ID "
+                + "WHERE df.DIRECTOR_ID = ?"
```
Здорово запрос построили, приятно читать

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java:
```diff
@@ -101,25 +121,59 @@ public Map<String, Object> filmToMap(Film film) {
+                        "WHERE F.FILM_ID IN (" +
+                        "SELECT FILM_ID FROM LIKES " +
+                        "WHERE USER_ID IN (" +
+                        "SELECT L1.USER_ID FROM LIKES L1 " +
```
А здесь подзапрос не очень читается, его бы как то табом сдвинуть вправо

EndorpfinSplash replied:
Запросы с оператором IN обычно могут вызывать проблемы, когда данных много становится. 
Я бы переписал на JOIN это

VladimirTitov4 replied:
Проблем не должно возникнуть, разве что в этом спринте запрещено использовать более 1 подзапроса, сразу не заметил. 

Ну и да, в любом случае JOIN лучше

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmReviewDbStorage.java:
```diff
@@ -0,0 +1,146 @@
+    public Optional<Review> getReviewById(Integer reviewId) {
+        final String SQL_GET_REVIEW = "select * from FILM_REVIEW where id = ?";
+        try {
+            Review review = jdbcTemplate.queryForObject(SQL_GET_REVIEW, reviewRowMapper, reviewId);
```
Отлично, что используете именно queryForObject

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/SearchDbStorage.java:
```diff
@@ -0,0 +1,51 @@
+                "LEFT JOIN DIRECTOR_FILM DF on f.FILM_ID = DF.FILM_ID " +
+                "LEFT JOIN DIRECTORS D on D.DIRECTOR_ID = DF.DIRECTOR_ID " +
+                "LEFT JOIN LIKES L on f.FILM_ID = L.FILM_ID WHERE (f.NAME ilike concat('%', ?, '%')) " +
+                "OR (d.NAME ilike concat('%', ?,'%')) " +
```
Плюс за ilike

EndorpfinSplash replied:
Запросы с оператором OR могут сильно снижать производительность запроса, лучше их избегать по возможности, например так:
WHERE instr(f.NAME || d.NAME, ?) >0

VladimirTitov4 replied:
Не думаю, что сильно снижают, разве что немного

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -54,12 +81,77 @@ public Film deleteLike(Integer filmId, Integer userId) {
+                film -> {
+                    film.getLikes().addAll(likeStorage.getLikesById(film.getId()));
+                    film.getGenres().addAll(genreStorage.getGenreListById(film.getId()));
+                    film.setDirectors(directorStorage.getDirectorsListById(film.getId()));
```
Здесь запосы в БД в цикле. 
Нужно заранее получить все сущности и накопить их в HashMap, которую уже потом итерировать. 
Используйте stream(collect(toMap .... )

Это точно нужно поправить, т.к. запросы в БД в цикле - очень плохо

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -54,12 +81,77 @@ public Film deleteLike(Integer filmId, Integer userId) {
+        }
+        List<Film> films = filmStorage.getSortedFilms(id, sort);
+        films.forEach(film -> {
+            film.setDirectors(directorStorage.getDirectorsListById(film.getId()));
```
Так же 3 запроса в БД в цикле

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -54,12 +81,77 @@ public Film deleteLike(Integer filmId, Integer userId) {
+        userStorage.findUserById(friendId);
+        List<Film> commonFilms = filmStorage.getCommonFilms(userId, friendId);
+        commonFilms.forEach(film -> {
+            film.setDirectors(directorStorage.getDirectorsListById(film.getId()));
```
цикл в бд

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -54,12 +81,77 @@ public Film deleteLike(Integer filmId, Integer userId) {
+    public List<Film> getPopularFilmsByGenreAndYear(Integer count, Integer genreId, Integer year) {
+        List<Film> films = filmStorage.getPopularFilmsByGenreAndYear(count, genreId, year);
+        for (Film film : films) {
+            film.setGenres(genreStorage.getGenreListById(film.getId()));
```
цикл в бд

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -40,26 +52,51 @@ public User findUserById(Integer userId) {
+        findUserById(userId);
+        List<Film> films = filmStorage.getRecommendations(userId);
+        for (Film film : films) {
+            film.setGenres(genreStorage.getGenreListById(film.getId()));
```
цикл в бд

VladimirTitov4 replied:
Не исправлено

EndorpfinSplash committed:
perf: avoid multiple DB requests example

Commit:
perf: getFilmLikesMap created

Commit:
perf: getFilmDirectorsMap created + final refactoring

Commit:
style: sense variables involved

IliaKanyshev committed:
checkstyle : whitespaces

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/GenreDbStorage.java:
```diff
@@ -38,14 +43,58 @@ public List<Genre> getGenresList() {
+                    .id(filmGenreSet.getInt("genre_id"))
+                    .name(filmGenreSet.getString("name"))
+                    .build();
+            filmGenresMap.computeIfAbsent(filmId, flmId -> new ArrayList<>()).add(genre);
```
А у вас в мапу только один раз filmId попадает ? 
Я к тому, что вы делаете computeIfAbsent, а что если present ? Нужно же в существующий список добавить значение

EndorpfinSplash replied:
если present, то добавится в существующий список.

VladimirTitov4 replied:
Да, отлично реализовали.

Commit:
style: fix
fix: getRecommendations enreachment.
perf: Optimize query for getFilmListByTitleAndDirector
